### ISSUE TITLE
DISCUSS: future of docopt-ng

### ISSUE THREAD

**Issue by NickCrews** [jazzband/docopt-ng#54]
This is a broad topic that I've been thinking about recently, that has come up in several places, such as #4 and #45

What does the community want for the future of docopt-ng?
1. A maintained and type-hinted port of the original docopt
2. that, plus more features
3. something else?

I have so far been pursuing 1. This was because
- I was estimating that most users are just porting over from the old docopt to here.
- There are now many more CLI libraries than when docopt was released, so I think any new applications are more likely to use one of the more modern frameworks, that provides type hints, input validation etc. Docopt inherently is limited in those two aspects because it operates on plain docstrings, so those would have to be done as a post-processing step, [leading to code duplication](https://github.com/jazzband/docopt-ng/pull/4). Perhaps we should add this to the README so that new users know the limitations coming in.
- I liked the scope that @keleshev started with, `a function that takes a usage message and sys.argv, and returns a dict`. It's well defined, has a narrow interface but a complex implementation, and composes well. In release [0.9.0](https://github.com/jazzband/docopt-ng/releases/tag/0.9.0) I even removed the `magic` stuff in pursuit of this (see the changelog for rationale)
- So far, additional features have been able to be implemented as extensions, rather than needing to change this repo
   - the magic stuff could have been an extension
   - [`docopt-dispatch`](https://github.com/jazzband/docopt-ng/issues/45) could be implemented as an extension
- I have limited time and already am not giving this all the attention that it should probably be getting, so a smaller scope is more tenable for me.

I would love to hear people's opinions here. I don't have a lot of open source maintainership experience, so I apologize for any frustration that is caused by that.

Cheers,
Nick

**Comment by Sylvain303** [comment]
Hello @NickCrews,

I'm the maintainer of `docopts` with an `S` for shell in bash. Written in go. I'm in a huge pause on this project since last September.

My own plan was to totally rewrite the golang parser to fully handle parse errors of docopt language, hand have a really nice diagnostic behavior. AST and much more. This is an handcrafted parser and lexer. It's still at some PoC level.

I appreciate the original goal of docstring => to dict. In bash it makes even more sense. Then I handle the argument the way I want, in python too. If my docopt parser finally succeed to have a really nice error reporting (better achieved in handcrafted parser than in generated one) I was planing to port it back to python as well. The original docopt project is frozen but full of really nice issues (about my goal on error reporting).

Error reporting / diagnostic is mostly the only thing I need for a cli at my side.

Hope that helps.
Sylvain.

**Comment by NickCrews** [comment]
Hi @Sylvain303, nice to meet you!

If the multiple ports were able to share one grammar, I think that would help unify the whole ecosystem, and reduce bugs. If someone wrote that grammar, would you want to use it? I'm no expert in grammars so I don't know which one we would want (EBNF, PEG, etc?). I think the biggest drawback would be the generic errors, as you say. IDK how to address this, maybe in each language's grammar-parsing library there would be a way to catch the parsing errors and wrap them in more helpful error messages?

**Comment by Sylvain303** [comment]
> If the multiple ports were able to share one grammar, I think that would help unify the whole ecosystem, and reduce bugs.

Basically that's what was done. On the code side, not a well defined grammar. Then the python development stopped. The python original docopt was mostly "blindly" ported into other some language. That was the case in golang and some other I looked at. The parser algorithm was almost verbatim in Go the same as in the original python one.

This issue discuss about grammar: https://github.com/docopt/docopt/issues/330 (pointer to and language description used in F# port)

I made an analysis myself here: https://github.com/docopt/docopts/blob/dev-grammar/grammar/docopt_language.ebnf

> If someone wrote that grammar, would you want to use it? I'm no expert in grammars so I don't know which one we would want (EBNF, PEG, etc?). I think the biggest drawback would be the generic errors, as you say.

I'm not a grammar parser expert myself, but I love it since my first met of the GNU lex & Yacc book 20 years ago...

The docopt language is not trivial (on grammar definition) and has ambiguity. I needed a state lexer to tokenise it the way I wanted.  

What the book was saying, is that grammar generator greatly simplify the complexity of writing a parser and its maintenance. But the error handling is mostly limited. I tried some generator and I was limited. I finally looked at golang owns parser which is a dedicated code and lexer too, and decided to write my own for docopt too.

F# port was using some F# language trick which is friendly for parsing and matching rules. Other language has some similar facilities like in Rust I think.

>  would you want to use it?

I'm actually using the original golang one in `doctops`. This one also stopped, as mentioned it was a port. Go has now some interesting argument parser, that also adresses the command line completion. As in `kubectl` for example. Still docopt is the only one based on documentation parsing avoiding to remember the object side of other argument parser and their fixed model. As  @keleshev  was mentioning back in his original video presentation in 2012 at PyCon.

Yes I would use a shared grammar / parser and contribute to it. Obviously it need to satisfy my own need on error reporting too. And we may also need to agree on some behavior fix on some strange docopt original behavior (ex: dict mapping change when `Options:` is present with argument `--long=<argument_for_long>`) 

> IDK how to address this, maybe in each language's grammar-parsing library there would be a way to catch the parsing errors and wrap them in more helpful error messages?

I turn the problem in this way: *what is the result I would like to have with error handling?*
Then I'm experimenting with the parser and erroneous ambiguous docopt input.

When I would have the AST, obviously I will also generate the command line completion too. :wink:

**Comment by FelixSchwarz** [comment]
Just a random Python developer: I always loved docopt but was worried due to the lack of updates. I considered switching away from docopt but now that I found this project I hope I can use it for much longer (also switching away from docopt would take a long time as we have hundreds of scripts using docopt).

The thing I always liked about docopt was that the file had a easily readable help text at the top and that text was in sync with the actual script. I don't need more features but better error messages and diagnostics would be nice.
Over the years I encountered some parsing bugs/shortcomings in docopt but I was able to work around these so I don't remember any specifics anymore.

**Comment by Cyleigh** [comment]
A colleague suggested docopt as a solution to a problem that I'm currently having.  I'm building a library of command line tools, and templates about how to use the tool.  Currently, there is a manual process of entering the command line with all its arguments etc, which is obviously not ideal.  I'd like docopt to spit out my list of arguments with the name, flag, any choice options and the help text.  This is a small part of a much larger project, so I was definitely hoping to find something already in the wild but haven't quite found what I need.

**Comment by NickCrews** [comment]
@Cyleigh that use case makes sense, thanks for the thoughts! Does docopt satisfy your needs, or is there something missing?

**Comment by Cyleigh** [comment]
> @Cyleigh that use case makes sense, thanks for the thoughts! Does docopt satisfy your needs, or is there something missing?

I'm not necessarily the most experienced python user, so I am having some trouble.  For example, the help string of options and arguments are not returned in ParsedOptions, and it very much assumes that I will want to parse a command line (which I don't want to do on creation).

**Comment by sebastian-philipp** [comment]
I continue to really enjoy docopt: it gives me a CLI without the need to thing about any APIs.

**Comment by jfolz** [comment]
Just found this and thought I'd chime in. I'm using docopt-ng in my project [sqltrack](https://sqltrack.readthedocs.io). Specifically, I needed to parse arguments without actually running the code, and put those arguments into a database. Think scheduling a job on an HPC cluster, where the software available on the head node is very limited. At least back then, I found no other option besides putting the argument parsing into a separate file that doesn't depend on any third-party libraries.
In practice, missing data types were a problem, but I found a solution that worked pretty well: [Just guess](https://sqltrack.readthedocs.io/en/stable/docopt.html#argument-types). There are a number of rules based on suffix matching and users can add their own. E.g., an option that matches `*-path` would be converted to `pathlib.Path`. If no rule applies cleanly, it tries `int`, `float`, `complex`, and finally `json.loads`.
Overall, I'm pretty happy how this turned out. Argument parsing using docopt-ng is opt-in, and I haven't had any complaints about this scheme.
In conclusion, I'm happy to use docopt-ng as-is, and would appreciate if it remains functional with future Python versions :)

**Comment by dwt** [comment]
Having used the original docopt - so excuse me if you already support this.

I would really love for a special argument that just dumps a structured (json / yaml) representation of the argument parser, that easily allows me to check what is parsed and wether changes I did might have changed the way args will be parsed.

This could also allow diffing between the way stuff is parsed in versions (as a kind of acceptance tests).

That is something I always missed when using the original docopt, that I always just used very rarely and was having a lot of trouble remembering the intricacies of the docopt configuration language.

**Comment by NickCrews** [comment]
@dwt that feature makes sense, and should be theoretically be possible. 

If you want to discuss this further, please make a new feature request issue, so this thread doesn't go off topic. But in short:
I am not going to implement that (don't have the time time or interest), but I would consider merging a PR if
- I didn't have to do huge amounts of edits/review on it
- no additional dependencies, or very lightweight ones
- didn't look like a huge maintenance burden. I'm skeptical this would be possible
- didn't break existing users
I don't know of any common json schema that would work for this. I would love to not reinvent the wheel here, so would be worth it to explore what other. It also seems to me like "checking for a change in parsing behavior" would not be that useful on its own. If the behavior started off incorrect, you aren't actually catching that. It seems more useful to actually test `assert my_parse("myprogram new --name foo") == <some dict>`, even if it might be a bit more verbose. Also would serve as documentation.

**Comment by Sylvain303** [comment]
> Having used the original docopt - so excuse me if you already support this.
> 
> I would really love for a special argument that just dumps a structured (json / yaml) representation of the argument parser, that easily allows me to check what is parsed and wether changes I did might have changed the way args will be parsed.
> 
> This could also allow diffing between the way stuff is parsed in versions (as a kind of acceptance tests).
> 
> That is something I always missed when using the original docopt, that I always just used very rarely and was having a lot of trouble remembering the intricacies of the docopt configuration language.

@dwt 

Hello, I'm not very active yet on `docopts` coding (golang version for bash) watch-out the extra 'S'. But I
try to implement such thing:  dumping what was actually parsed by docopt parser.

Could you create an new issue with some sample of what you would love to see outputted? That would be very useful. 🤩

**Comment by dwt** [comment]
@Sylvain303 @NickCrews Thanks for the feedback, I have opened a separate bug at #61

**Comment by nkakouros** [comment]
I am a heavy user of docopt. The selling points are, I think, three:

- it helps keep documentation and code in sync,
- it is a lightweight and straightforward dependency,
- it supports a familiar and intuitive help text syntax.

Personally, I would like for docopt to be maintained so that it follows newer versions of Python and its quality improves (there are reported bugs and ambiguities). 

Regarding new features, it would be great to have them if they contribute to the selling points without compromising any of them. Given that there is an ecosystem of other projects around docopt, it would also be nice to maintain an up-to-date list of such projects.

**Comment by sebastian-philipp** [comment]
one thing that bothers me is that the API of argparse is super unintuitive, which is pretty much solved by docopt.

**Comment by frnhr** [comment]
Another random dev here, I keep returning to docopt. 

Alternatives are just so much less intuitive and feel bulky. If I need validation, I don't mind a bit of duplication, for the benefits of KISS and the documentation-first approach.

I wish there was a sleek way of combining with pydantic (without duplication), though.

But definitely don't want to see docopt gone!

**Comment by Notmarrco** [comment]
Same here, I like the simplicity of docopt, and I wasn't even annoyed by docopt being unmaintained, but now I'll switch to docopt-ng.
I use it with [commandopt](https://github.com/jaegerbobomb/commandopt) for all my projects (disclaimer : I wrote it) to easily execute functions (commands) given docopts arguments.

The only thing I would dream to see is an autocomplete, so perhaps some external tool taking a docopt and turning it into bash completions, but imo that's out of the scope of docopt-ng.

**Comment by jbaber** [comment]
A maintained, tested, typehinted docopt that continues to work with updates to Python makes me very happy.  A pure __doc__ to dict converter, too, would be fine, too.  I'm happy to parse the arguments myself.  In particular, it makes parsing with custom functions an option.

**Comment by Solganis** [comment]
Hi! I made a fork of our beloved docopt - [docopt2](https://github.com/Solganis/docopt2).

On the duplication worry @NickCrews raised: it doesn't have to be a second declaration. `docopt2 stub prog.py` generates the dataclass *from* the usage message, so the docstring stays the only place the arguments are written down. Then `docopt(__doc__, schema=Args)` gives you `args.speed` as an `int` under `mypy --strict` instead of `Any`.

There is rustc-style error output and a usage linter in there too. Would love to hear what you think.

**Comment by Sylvain303** [comment]
@Solganis I was the maintainer of `docopts` as mentioned [above](https://github.com/jazzband/docopt-ng/issues/54#issuecomment-1948074274). 

I'm actually far away from the project.

I plan to abandon the role and the project too. Because I don't have time anymore. And also because of the model in bash side of the `eval docopts "Docopt string to parse"` which is absolutely unsafe from my sysadmin security biases eyes 😱. I'm also opposed of the `curl-pipe-sudo-bash` install new  *absolutely unsafe* strange *standard* of way of distributing software. (Which is another story, I agree)

From my longtime running and following docopt (since back in 2015), I would say for other language than bash, and especially python it as very nice and hacky way (I love it) to both render option parsing and self documentation. 😁

Unfortunately the parser is a complicated computer programming stuff, and also the description language for Docopt language is a bit odd to parse. I dept in the original's python parsing code (Golang one was exactly the same) and did start a full rewrite PoC in Go many years ago. Especially to grab the AST and odd Docopt language features, that seemed buggy to me. Like the change from `-F <option>` to `--long=<option>` data structure result. Which is perfectly explainable when you know the parser, but quite odd for the casual user.

I waited 2 years to be granted to commit and manage in the main sub-repository by contacting some granted other people. The original creator has made his way elsewhere and did not respond.

My advice here would be, try to contact all the people still being part of the project original docopt to try to regain some grant privilege on the main domain github/docopt stuff. Probably some huge social effort required, but some issues still arrive in the main project, which is stalled since many years. 🤷🏻‍♂️

And find parsing expert and lover, as much of the derivative adapter in any language took the benefits, and the biases, to reproduce the original parsing  algorithm. The parser which lack useful error parsing messages of step to catch it. This strong basement on the main Docopt language and it's parser would be valuable for all Docopt sub-project based on a 1:1 parser reproduction.

I would love to see:

`doctop_helper explain 'Docopt input' -- my-action --why dont_you --eat=my-option "here?"'`

Good luck.

**Comment by Solganis** [comment]
Thanks @Sylvain303, that is generous coming from someone who actually lived inside the parser.

The `explain 'usage'` helper you are describing is mostly there already. 
The error messages you said the parser never had are what `docopt2 check` prints:

```
$ docopt2 check prog.py
warning: option `--nope` is declared but never used
   |
   |  in the options:
   |      --nope  Unused.
   |      ^^^^^^ declared here
   |
   = help: add `--nope` to a usage line, or add `[options]` to accept it
```

And for "what does this usage actually accept", `docopt2 examples` walks the grammar and prints real invocations.   
`parse_tree(doc)` hands the AST back as data, which is what @dwt was after in #61.

On regaining access to the docopt org: I hear you, but I have decided not to go that way.   
If a signpost comes back over there, it should come from the org, not from a fork author pointing at his own repo.

Either way, thank you for docopts, and good luck with what is next.

**Comment by NickCrews** [comment]
I'm glad you made a fork, I think that is the right path for that level of exploration/innovation. If, during that process, you discover ways to improve docopt-ng that are backwards compatible, then feel free to file an issue :)
